### PR TITLE
Adds plot auto-detection.

### PR DESCRIPTION
**Pull request by boehmseb**
Adds an automtated plot registry (similar to project registry) so creating a new plot no longer involves manually creating an entry in the list of available plots.

resolves se-passau/VaRA#549

### CODE REVIEW

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VaRA/VaRA-Tool-Suite/pullRequest?prid=4957273&bid=16381693)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VaRA/VaRA-Tool-Suite/pullRequest?prid=4957273&bid=16381693)

File: varats/paper/case_study.py [hunk 1]
```diff
@@ -898,9 +899,9 @@ def extend_with_smooth_revs(case_study: CaseStudy, cmap: CommitMap,
     new_revisions = plot.calc_missing_revisions(boundary_gradient)
 
     # Remove revision that are already present in another stage.
-    new_revisions = [
+    new_revisions = set([
```
Comment (vulder) [inline varats/paper/case_study.py]:
```suggestion
    new_revisions = {
```

File: varats/paper/case_study.py [hunk 2]
```diff
@@ -898,9 +899,9 @@ def extend_with_smooth_revs(case_study: CaseStudy, cmap: CommitMap,
...
         rev for rev in new_revisions if not case_study.has_revision(rev)
-    ]
+    ])
```
Comment (vulder) [inline varats/paper/case_study.py]:
```suggestion
    }
```

File: varats/paper/case_study.py (hunk 1 above)
Comment (vulder) [reply varats/paper/case_study.py]: using python set comprehension

File: varats/plots/plots.py [hunk 3]
```diff
@@ -1,45 +1,55 @@
...
-from varats.plots.blame_lorenz_curve import (BlameLorenzCurve,
-                                             BlameGiniOverTime)
-from varats.plots.repository_churn import RepoChurnPlot
-from varats.plots.plot import Plot
+
+# avoid circular import
+# see https://www.stefaanlippens.net/circular-imports-type-hints-python.html
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from varats.plots.plot import Plot
```
Comment (vulder) [inline varats/plots/plots.py:13]: that's nice, did not know that :D

**Review by vulder — CHANGES_REQUESTED** [review]
could you please also check if we can fix the naming problem in the help string for vara-plots with the new metaclass.
At the moment, vara-plot prints strange strings als options for the plot type. Can we use the new NAME attr there?

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VaRA/VaRA-Tool-Suite/pullRequest?prid=4957273&bid=16381693)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/plots/plots.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VaRA/VaRA-Tool-Suite/pullRequest?prid=4957273&bid=16381693)

**Comment by boehmseb** [comment]
> could you please also check if we can fix the naming problem in the help string for vara-plots with the new metaclass.
> At the moment, vara-plot prints strange strings als options for the plot type. Can we use the new NAME attr there?

See my latest commit.

**Comment by vulder** [comment]
> > could you please also check if we can fix the naming problem in the help string for vara-plots with the new metaclass.
> > At the moment, vara-plot prints strange strings als options for the plot type. Can we use the new NAME attr there?
> 
> See my latest commit.
This only shows if the user types it wrong, is there a way to show the list of all plots in the `--help` menu of `vara-plot`? In the old version this only printed a list of enumish like things but no clear plot names.

**Review by vulder — CHANGES_REQUESTED** [review]
looks good except for the `--help` issue, do you think we can add this?

If not we can also merge it this way

**Comment by boehmseb** [comment]
Should be possible. I can look into it later today.

**Comment by boehmseb** [comment]
This commit adds a listing of available plot types to the output of `vara-plot -h`.

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VaRA/VaRA-Tool-Suite/pullRequest?prid=4957273&bid=16381693)

**Review by vulder — APPROVED** [review]
nice, this looks good now :smile:

**Pull request merged**